src/core/public/context/context.ts

```diff
@@ -0,0 +1,281 @@
+   * @returns The `ContextContainer` for method chaining.
+   */
+  registerContext<TContextName extends keyof TContext>(
+    plugin: string | CoreId,
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, THandlerParameters>
+  ): this;
+
+  /**
+   * Create a new handler function pre-wired to context for the plugin.
+   *
+   * @param plugin - The plugin ID for the plugin that registers this context.
+   * @param handler
+   * @returns A function that takes `THandlerParameters`, calls `handler` with a new context, and returns a Promise of
+   * the `handler` return value.
+   */
+  createHandler(
+    plugin: string | CoreId,
+    handler: Handler<TContext, THandlerReturn, THandlerParameters>
+  ): (...rest: THandlerParameters) => Promisify<THandlerReturn>;
+}
+
+type ContextSource = PluginName | CoreId;
+
+/** @internal */
+export class ContextContainerImplementation<
+  TContext extends {},
+  THandlerReturn,
+  THandlerParameters extends any[] = []
+> implements ContextContainer<TContext, THandlerReturn, THandlerParameters> {
+  /**
+   * Used to map contexts to their providers and associated plugin. In registration order which is tightly coupled to
+   * plugin load order.
+   */
+  private readonly contextProviders = new Map<
+    keyof TContext,
+    {
+      provider: ContextProvider<TContext, keyof TContext, THandlerParameters>;
+      source: ContextSource;
+    }
+  >();
+  /** Used to keep track of which plugins registered which contexts for dependency resolution. */
+  private readonly contextNamesBySource: Map<ContextSource, Array<keyof TContext>>;
+
+  /**
+   * @param pluginDependencies - A map of plugins to an array of their dependencies.
+   */
+  constructor(
+    private readonly pluginDependencies: ReadonlyMap<PluginName, PluginName[]>,
+    private readonly coreId: CoreId
+  ) {
+    this.contextNamesBySource = new Map<ContextSource, Array<keyof TContext>>([[coreId, []]]);
+  }
+
+  public registerContext = <TContextName extends keyof TContext>(
+    source: ContextSource,
+    contextName: TContextName,
+    provider: ContextProvider<TContext, TContextName, THandlerParameters>
+  ): this => {
+    if (this.contextProviders.has(contextName)) {
+      throw new Error(`Context provider for ${contextName} has already been registered.`);
+    }
+
+    if (typeof source === 'symbol' && source !== this.coreId) {
+      throw new Error(`Symbol only allowed for core services`);
+    }
+
+    this.contextProviders.set(contextName, { provider, source });
+    this.contextNamesBySource.set(source, [
+      ...(this.contextNamesBySource.get(source) || []),
+      contextName,
+    ]);
+
+    return this;
+  };
+
+  public createHandler = (
+    source: ContextSource,
+    handler: Handler<TContext, THandlerReturn, THandlerParameters>
+  ) => {
+    if (typeof source === 'symbol' && source !== this.coreId) {
+      throw new Error(`Symbol only allowed for core services`);
```

Comment:
btw why not? then plugins can create uniq id by itself (if don't use `Symbol.for('...')`, ofc.)

Reply:
Because this id has to be the same id as the kibana.json file so we can map it to the dependencies for this plugin.